### PR TITLE
[MLX] Fix test_batched_decode_matches_solo failing past EOS

### LINKED ISSUE #32441: [Bug] [MLX] test_batched_decode_matches_solo asserts bitwise solo/batched token equality, which no batch-shape-varying backend can provide
## Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Describe the bug

`test_mlx_reference_correctness.py::TestMlxReferenceCorrectness::test_batched_decode_matches_solo` fails on unmodified `main` (`e14068d`) on Apple Silicon.

The test decodes a fixed `BATCH_HORIZON = 24` steps and asserts that a request's token stream is identical whether it is decoded alone or inside a `decode_batch`. That assertion cannot hold: bitwise-identical logits across two different batch shapes is not an achievable property. The failure is in the test's premise, not in the MLX cache/slot bookkeeping it is meant to guard — that bookkeeping is correct, and I verified it stays correct below.

This test is registered on `stage-b-e2e-mlx` (and on `base-a-test-cpu`, where it skips for lack of `mlx`), so it will turn the Apple Silicon lane red as soon as the self-hosted macOS runners from the [roadmap](https://github.com/sgl-project/sglang/issues/19137) come online.

## Reproduction

```
python3 -m pytest test/registered/unit/hardware_backend/mlx/test_mlx_reference_correctness.py -q
```

```
FAILED test_batched_decode_matches_solo
First differing element 6:
151645
151643
  prompt: 'What is the capital of France? Answer in one word.'
  first divergence @ index 6 (len ref=24 sgl=24)
  ref text: 'Paris.<|im_end|>\n<|endoftext|>\n<|endoftext|>\n<|im_end|>\n<|im_start|>\n...'
  sgl text: 'Paris.<|im_end|>\n<|endoftext|>\n<|im_end|>\n<|im_end|>\n<|im_start|>\n...'
```

Note where the divergence sits: index 6, four tokens *after* the model already emitted EOS (`<|im_end|>`, index 2). The answer itself — `Paris.` — is identical.

`test_greedy_matches_reference_exact` in the same file passes. It stops at EOS.

## Root cause

Three experiments on the same machine, driving `MlxModelRunner` directly and capturing full logits per step rather than just the argmax.

**1. The runner is deterministic for a fixed batch shape.** Decoding the same request solo twice:

```
step1..5: max|dlogit| = 0.00000   (bitwise identical)
```

**2. The perturbation comes from the batch shape, not from padding.** Solo (`B=1`) vs a batch of two *equal-length* prompts, which takes neither the padding nor the mask path (`needs_padding=False`, `pad=[0,0]`):

```
step1: max|dlogit|=0.02734  rel=1.16e-03
step2: max|dlogit|=0.02734  rel=1.05e-03
step3: max|dlogit|=0.03320  rel=1.01e-03
step4: max|dlogit|=0.02686  rel=1.19e-03
step5: max|dlogit|=0.03906  rel=2.16e-03
```

**3. The ragged-padding + mask path adds exactly zero error.** Same comparison against the three-prompt ragged batch that the test actually builds (`needs_padding=True`, `seq_lens=[30, 30, 26]`, `pad=[0, 0, 4]`):

```
step1: max|dlogit|=0.02734  rel=1.16e-03
step2: max|dlogit|=0.02734  rel=1.05e-03
step3: max|dlogit|=0.03320  rel=1.01e-03
step4: max|dlogit|=0.02686  rel=1.19e-03
step5: max|dlogit|=0.03906  rel=2.16e-03
```

Byte-for-byte the same numbers as experiment 2. So `BatchedDecodeContext`'s `valid_lens`/`pad_sizes` and the `-inf` attention mask in `MLXAttentionWrapper` are exact — they contribute no error at all. What remains is fp16 GEMM reduction order changing with the batch dimension, which is inherent to batched inference.

**Why a ~1e-3 perturbation flips a token.** Per-step top-2 logit margin for the failing request:

| step | max&#124;dlogit&#124; | solo top-2 margin | batched top-2 margin | argmax |
|---|---|---|---|---|
| 1 | 0.027 | 1.547 | 1.547 | same |
| 2 | 0.027 | 11.039 | 11.039 | same |
| 3 | 0.033 | 13.953 | 13.953 | same |
| 4 | 0.027 | 2.328 | 2.328 | same |
| 5 | 0.039 | 0.844 | 0.844 | same |
| **6** | **1.273** | 0.789 | **0.016** | **flip** |

Before EOS the margin is 1.5–14, orders of magnitude above the perturbation. Past EOS the model is off-distribution, emitting `<|endoftext|>`/`<|im_start|>` filler, and the margin collapses to 0.016 — at which point fp16 noise decides the argmax. (The jump in `max|dlogit|` at step 6 is the MoE router: near-tied gate scores make top-k expert selection discontinuous, so a 1e-3 input perturbation produces an O(1) output change.)

**Divergence is always strictly after EOS.** Across all three prompts the test uses:

| case | first EOS | first divergence | prompt |
|---|---|---|---|
| 0 | none in 24 | none — identical | `List the first 10 prime numbers, comma separated.` |
| 1 | index 2 | index 6 | `What is the capital of France? Answer in one word.` |
| 2 | none in 24 | none — identical | `Write one short sentence about the ocean.` |

The two prompts that never reach EOS match for all 24 steps. Only the one that terminates early diverges, and only in the filler past its terminator. Up to and including EOS, all three match exactly — the isolation property the test exists to check does hold.

## Proposed fix

Bound the comparison at the first EOS, matching what `test_greedy_matches_reference_exact` already does, and document why bitwise equality past EOS is not a property the backend can offer.

This does not weaken the guard. With the fix applied, three injected regressions in `MLXAttentionWrapper` / `BatchedDecodeContext` each still turn the test red:

| mutation | fixed test |
|---|---|
| baseline (no mutation) | passes |
| padding mask disabled (`needs_padding` → `False`) | **fails** |
| `valid_lens` off by one (`offsets + 1` → `offsets`) | **fails** |
| cross-request bleed (reverse per-layer cache order) | **fails** |

The third is precisely the "one request's state bleeds into another's" scenario the test was written for.

PR: #32442

## Environment

```
Python: 3.11.15
macOS: 26.5.2
Chip: Apple M5 Pro (15 cores)
Unified Memory: 48 GB
max_recommended_working_set_size: 37.4 GB
sglang: main @ e14068d
mlx: 0.32.0
mlx-lm: 0.31.3
mlx-metal: 0.32.0
torch: 2.11.0
transformers: 5.12.1
Model: mlx-community/Qwen1.5-MoE-A2.7B-Chat-4bit (test default)
```

Full MLX unit suite on this machine at `e14068d`: `2 failed, 85 passed, 19 skipped`. The other failure is the already-tracked #30550. With this fix: `1 failed, 86 passed, 19 skipped`.

cc @LarrySimingDeng (added the test in #29440) @jlee5814 @yeahdongcn @jonahbernard

### PR DESCRIPTION
**Pull request by xiaolin2004**
## Motivation

Fixes #32441.

`test_mlx_reference_correctness.py::test_batched_decode_matches_solo` fails on unmodified `main` (`e14068d`) on Apple Silicon. It decodes a fixed 24-step horizon and asserts a request's tokens are identical decoded solo vs. inside a `decode_batch`. That premise does not hold: bitwise-identical logits across two different batch shapes is not an achievable property of any batched backend.

The bookkeeping the test guards is fine. The test is over-strict, and it will redden the Apple Silicon lane the moment the self-hosted macOS runners land.

Observed failure — note the divergence is four tokens *past* EOS, while the answer itself is identical:

```
First differing element 6: 151645 != 151643
  prompt: 'What is the capital of France? Answer in one word.'
  ref text: 'Paris.<|im_end|>\n<|endoftext|>\n<|endoftext|>\n<|im_end|>\n...'
  sgl text: 'Paris.<|im_end|>\n<|endoftext|>\n<|im_end|>\n<|im_end|>\n...'
```

`test_greedy_matches_reference_exact` in the same file passes — it stops at EOS.

## Root cause

Three experiments driving `MlxModelRunner` directly and capturing full logits per step (M5 Pro, `Qwen1.5-MoE-A2.7B-Chat-4bit`):

| comparison | padding/mask active? | max&#124;dlogit&#124; over 5 steps |
|---|---|---|
| solo vs solo, same shape | n/a | **0.00000** — bitwise identical |
| solo vs **equal-length** batch | no | 0.027, 0.027, 0.033, 0.027, 0.039 |
| solo vs **ragged** batch (`seq_lens=[30,30,26]`, `pad=[0,0,4]`) | yes | 0.027, 0.027, 0.033, 0.027, 0.039 |

Rows 2 and 3 are byte-for-byte identical, so `BatchedDecodeContext.valid_lens`/`pad_sizes` and the `-inf` mask in `MLXAttentionWrapper` contribute **exactly zero** error — that path is exact. The residual ~1e-3 relative delta is fp16 GEMM reduction order changing with the batch dimension, which is inherent to batched inference.

Why that flips a token only past EOS — top-2 logit margin for the failing request:

| step | max&#124;dlogit&#124; | solo margin | batched margin | argmax |
|---|---|---|---|---|
| 1–5 | 0.027–0.039 | 0.84 – 13.95 | 0.84 – 13.95 | same |
| **6** | **1.273** | 0.789 | **0.016** | **flip** |

Before EOS the margin is orders of magnitude above the perturbation. Past EOS the model emits control-token filler with a 0.016 margin, and fp16 noise decides the argmax. (The step-6 jump in `max|dlogit|` is the MoE router: near-tied gate scores make top-k expert selection discontinuous.)

Across all three prompts, divergence is *always* strictly after EOS — the two that never reach EOS match for all 24 steps:

| case | first EOS | first divergence |
|---|---|---|
| 0 `List the first 10 prime numbers…` | none in 24 | none — identical |
| 1 `What is the capital of France?…` | 2 | 6 |
| 2 `Write one short sentence about the ocean.` | none in 24 | none — identical |

## Modifications

One test file. Adds `_upto_eos()` and bounds the solo-vs-batched comparison at the first EOS, mirroring what `test_greedy_matches_reference_exact` already does. Docstrings record why bitwise equality past EOS is not on offer, with the measured numbers, so this doesn't get "fixed" back later.

No source changes — nothing in `srt/` is wrong here.

## Accuracy Tests

The guard is not weakened. With the fix applied, three injected regressions each still turn the test red:

| mutation in `MLXAttentionWrapper` / `BatchedDecodeContext` | fixed test |
|---|---|
| baseline, no mutation | passes |
| padding mask disabled (`needs_padding` → `False`) | **fails** |
| `valid_lens` off by one (`offsets + 1` → `offsets`) | **fails** |
| cross-request bleed (reverse per-request cache order) | **fails** |

The last is exactly the "one request's state bleeds into another's" scenario the test exists for.

Full MLX unit suite on Apple Silicon:

```
before:  2 failed, 85 passed, 19 skipped
after:   1 failed, 86 passed, 19 skipped
```

The remaining failure is the already-tracked #30550 (`test_metal_profiler`), untouched by this PR.

## Speed Tests and Profiling

N/A — test-only change.

## Environment

```
Python 3.11.15 · macOS 26.5.2 · Apple M5 Pro · 48 GB unified
mlx 0.32.0 · mlx-lm 0.31.3 · torch 2.11.0 · transformers 5.12.1
sglang main @ e14068d
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests). *This PR is itself a unit-test fix; mutation results above show the assertion still catches real regressions.*
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). *Rationale captured in the test docstrings.*

cc @LarrySimingDeng (author of the test, #29440) @jlee5814 @yeahdongcn @jonahbernard

---
### CI States

Latest PR Test (Base): :x: [Run #30270794070](https://github.com/sgl-project/sglang/actions/runs/30270794070)
Latest PR Test (Extra): :x: [Run #30270794064](https://github.com/sgl-project/sglang/actions/runs/30270794064)

### CODE REVIEW

**Review by LarrySimingDeng — APPROVED** [review]
LGTM. This test is mine originally and the EOS bound is the right call, byte-exact argmax past EOS was never the point. `test_greedy_matches_reference_exact` already does the same. One ask: keep at least one prompt that never hits EOS within the horizon (cases 0 and 2 today), maybe assert it, so the full-horizon coverage doesn't quietly shrink later.

**Comment by now-ing** [comment]
The substance of this change landed on main via #34166 (2026-08-10): `test_batched_decode_matches_solo` now truncates both sides at the first EOS with the same rationale in the docstring (`_truncate_at_eos`), and the patch no longer applies cleanly against that region.

I re-ran the file on current main (Apple Silicon M4 Max, mlx 0.31.2 / mlx-lm 0.31.3): both tests pass. One caveat for completeness: with the only local model I had (Qwen2.5-0.5B dense), solo vs batched raw 24-step streams agree exactly, so the fix's load-bearing path (divergence past EOS on the MoE default model) isn't exercised on my machine â but the mechanism matches your analysis.

One piece of this PR is still worth salvaging: the `assertTrue(any(self.eos_ids.isdisjoint(seq) for seq in solo))` guard. It ensures at least one case runs the full horizon without EOS, so the isolation property can't silently shrink to comparing a couple of tokens if the prompt fixture changes later. Would you be open to rebasing this PR down to just that increment (or I can open a tiny follow-up)? After that this PR can close as superseded by #34166.
